core/src/pow/common.rs

```diff
@@ -151,3 +128,80 @@ macro_rules! to_edge {
 		T::from($n).ok_or(ErrorKind::IntegerCast)?
 	};
 }
+
+/// Utility struct to calculate commonly used Cuckoo parameters calculated
+/// from header, nonce, sizeshift, etc.
+pub struct CuckooParams<T>
+where
+	T: EdgeType,
+{
+	pub edge_bits: u8,
+	pub proof_size: usize,
+	pub num_edges: u64,
+	pub siphash_keys: [u64; 4],
+	pub easiness: T,
+	pub edge_mask: T,
+}
+
+impl<T> CuckooParams<T>
+where
+	T: EdgeType,
+{
+	/// Instantiates new params and calculate easiness, edge mask, etc
+	pub fn new(
+		edge_bits: u8,
+		proof_size: usize,
+		easiness_pct: u32,
+		cuckatoo: bool,
+	) -> Result<CuckooParams<T>, Error> {
+		let num_edges = 1 << edge_bits;
+		let num_nodes = 2 * num_edges as u64;
+		let easiness = if cuckatoo {
+			to_u64!(easiness_pct) * num_nodes / 100
+		} else {
+			to_u64!(easiness_pct) * num_edges / 100
+		};
+		let edge_mask = if cuckatoo {
+			to_edge!(num_edges - 1)
+		} else {
+			to_edge!(num_edges / 2 - 1)
+		};
```

Comment:
We're probably better off consolidating 'num_edges' between cuckoo and cuckatoo (I think cuckatoo29 = cuckoo 30)?